src/storage/v2/inmemory/storage.cpp

```diff
@@ -272,111 +272,299 @@ std::optional<VertexAccessor> InMemoryStorage::InMemoryAccessor::FindVertex(Gid
 }
 
 Result<std::optional<std::pair<VertexAccessor, std::vector<EdgeAccessor>>>>
 InMemoryStorage::InMemoryAccessor::DetachDeleteVertex(VertexAccessor *vertex) {
   using ReturnType = std::pair<VertexAccessor, std::vector<EdgeAccessor>>;
 
-  MG_ASSERT(vertex->transaction_ == &transaction_,
-            "VertexAccessor must be from the same transaction as the storage "
-            "accessor when deleting a vertex!");
-  auto *vertex_ptr = vertex->vertex_;
+  auto res = DetachDelete({vertex}, {}, true);
 
-  std::vector<std::tuple<EdgeTypeId, Vertex *, EdgeRef>> in_edges;
-  std::vector<std::tuple<EdgeTypeId, Vertex *, EdgeRef>> out_edges;
+  if (res.HasError()) {
+    return res.GetError();
+  }
 
-  {
-    std::lock_guard<utils::SpinLock> guard(vertex_ptr->lock);
+  auto &value = res.GetValue();
+  if (!value) {
+    return std::optional<ReturnType>{};
+  }
 
-    if (!PrepareForWrite(&transaction_, vertex_ptr)) return Error::SERIALIZATION_ERROR;
+  auto &[vertices, edges] = *value;
 
-    if (vertex_ptr->deleted) return std::optional<ReturnType>{};
+  MG_ASSERT(vertices.size() <= 1, "The number of detach deleted vertices is not less or equal to 1!");
 
-    in_edges = vertex_ptr->in_edges;
-    out_edges = vertex_ptr->out_edges;
-  }
+  return std::make_optional<ReturnType>(vertices[0], std::move(edges));
+}
+
+Result<std::optional<std::pair<std::vector<VertexAccessor>, std::vector<EdgeAccessor>>>>
+InMemoryStorage::InMemoryAccessor::DetachDelete(std::vector<VertexAccessor *> nodes, std::vector<EdgeAccessor *> edges,
+                                                bool detach) {
+  using ReturnType = std::pair<std::vector<VertexAccessor>, std::vector<EdgeAccessor>>;
 
+  // result of the operation
+  std::vector<VertexAccessor> deleted_vertices;
   std::vector<EdgeAccessor> deleted_edges;
-  for (const auto &item : in_edges) {
-    auto [edge_type, from_vertex, edge] = item;
-    EdgeAccessor e(edge, edge_type, from_vertex, vertex_ptr, &transaction_, &storage_->indices_,
-                   &storage_->constraints_, config_);
-    auto ret = DeleteEdge(&e);
-    if (ret.HasError()) {
-      MG_ASSERT(ret.GetError() == Error::SERIALIZATION_ERROR, "Invalid database state!");
-      return ret.GetError();
+  std::set<EdgeRef> deleted_edges_set;
+
+  // routine for deleting edges, please use this under the edge spin lock
+  auto mark_edge_as_deleted_func = [this](auto *edge_ptr) {
+    if (!edge_ptr->deleted) {
+      CreateAndLinkDelta(&transaction_, edge_ptr, Delta::RecreateObjectTag());
+      edge_ptr->deleted = true;
+      storage_->edge_count_.fetch_add(-1, std::memory_order_acq_rel);
+
+      // Need to inform the next CollectGarbage call that there are some
+      // non-transactional deletions that need to be collected
+      if (transaction_.storage_mode == StorageMode::IN_MEMORY_ANALYTICAL) {
+        auto *mem_storage = static_cast<InMemoryStorage *>(storage_);
+        mem_storage->gc_full_scan_edges_delete_ = true;
+      }
```

Comment:
Opportunity to defer the atomic write. 
Locally track, then do a single atomic write at end if needed.

Reply:
eg. 
```
  bool gc_needs_informing_edge_deletion = false;
  auto const inform_gc_edge_deletion = utils::OnScopeExit{[this, &gc_needs_informing_edge_deletion](){
    // Need to inform the next CollectGarbage call that there are some
    // non-transactional deletions that need to be collected
    if (gc_needs_informing_edge_deletion && transaction_.storage_mode == StorageMode::IN_MEMORY_ANALYTICAL) {
      auto *mem_storage = static_cast<InMemoryStorage *>(storage_);
      mem_storage->gc_full_scan_edges_delete_ = true;
    }
  }};
  
  bool gc_needs_informing_vertex_deletion = false;
  auto const inform_gc_vertex_deletion = utils::OnScopeExit{[this, &gc_needs_informing_vertex_deletion](){
    // Need to inform the next CollectGarbage call that there are some
    // non-transactional deletions that need to be collected
    if (gc_needs_informing_vertex_deletion && transaction_.storage_mode == StorageMode::IN_MEMORY_ANALYTICAL) {
      auto *mem_storage = static_cast<InMemoryStorage *>(storage_);
      mem_storage->gc_full_scan_vertices_delete_ = true;
    }
  }};
  ```
  and just set `gc_needs_informing_*_deletion`